libtransact/src/database/btree.rs

```diff
@@ -0,0 +1,737 @@
+    fn index_cursor(&self, index: &str) -> Result<DatabaseCursor, DatabaseError> {
+        let mut index_db = self
+            .db
+            .indexes
+            .get(index)
+            .ok_or_else(|| DatabaseError::ReaderError(format!("Not an index: {}", index)))?
+            .clone();
+
+        for transaction in self.transactions.iter() {
+            match transaction {
+                WriterTransaction::IndexPut {
+                    index: transaction_index,
+                    key,
+                    value,
+                } => {
+                    if index == transaction_index {
+                        index_db.insert(key.to_vec(), value.to_vec());
+                    }
+                }
+                WriterTransaction::IndexDelete {
+                    index: transaction_index,
+                    key,
+                } => {
+                    if index == transaction_index {
+                        index_db.remove(key);
+                    }
+                }
+                _ => (),
+            }
+        }
+
+        Ok(Box::new(BTreeDatabaseCursor::new(index_db)))
+    }
+
+    /// Returns the number of entries in the main database.
+    fn count(&self) -> Result<usize, DatabaseError> {
+        let count = self
+            .transactions
+            .iter()
+            .fold(0_i32, |acc, transaction| match transaction {
+                WriterTransaction::Put { .. } => acc + 1,
+                WriterTransaction::Delete { .. } => acc - 1,
+                _ => acc,
+            });
+        let total = self.db.main.len() as i32 + count;
+
+        Ok(total as usize)
+    }
+
+    /// Returns the number of entries in the given index
+    fn index_count(&self, index: &str) -> Result<usize, DatabaseError> {
+        let count = self
+            .transactions
+            .iter()
+            .fold(0_i32, |acc, transaction| match transaction {
+                WriterTransaction::IndexPut {
+                    index: transaction_index,
+                    ..
+                } => {
+                    if index == transaction_index {
+                        acc + 1
+                    } else {
+                        acc
+                    }
+                }
+                WriterTransaction::IndexDelete {
+                    index: transaction_index,
+                    ..
+                } => {
+                    if index == transaction_index {
+                        acc - 1
+                    } else {
+                        acc
+                    }
+                }
+                _ => acc,
+            });
+        let total = self
+            .db
+            .indexes
+            .get(index)
+            .ok_or_else(|| DatabaseError::ReaderError(format!("Not an index: {}", index)))?
+            .len() as i32
+            + count;
+
+        Ok(total as usize)
+    }
+}
+
+pub struct BTreeDatabaseCursor {
+    db: BTreeMap<Vec<u8>, Vec<u8>>,
```

Comment:
This should just wrap a DB reader - it also needs to exist in a read transaction.

Reply:
Or, at least, wrap a RwReadLock, like the Reader.

Reply:
Initially, I had BTreeDatabaseCursor take a reference to a btree map like below. 

```
pub struct BTreeDatabaseCursor<'a> {
    db: &'a BTreeMap<Vec<u8>, Vec<u8>>,
    iterator: Iter<'a, Vec<u8>, Vec<u8>>
}
```

 It worked fine for a cursor for the BTreeReader, but not for a cursor for the BTreeWriter, as I have to take into consideration the uncommitted transactions. The current implementation allows BTreeWriter to create a copy of the BTreeMap and apply the transactions without changing the original db.   Do you have suggestions on a better approach for this?

 Using a RwReadLock as you suggest, would not work to use  BTreeDatabaseCursor for the BTreeWriter implementation of DatabaseRead trait. 

I suppose I could create another struct to implement a cursor specific for BTreeWriter? With my current implementation, I was trying to avoid having to do that.